pkg/controllers/metrics/state/scraper.go

```diff
@@ -0,0 +1,121 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/aws/karpenter/pkg/controllers/state"
+	"knative.dev/pkg/logging"
+)
+
+const tickPeriodSeconds = 5
+
+type scraper interface {
+	getName() string
+	init(context.Context)
+	reset()
+	update(context.Context)
+}
+
+type MetricScraper struct {
+	Cluster *state.Cluster
+
+	scrapers         []scraper
+	terminateChan    chan struct{}
+	updateChan       chan struct{}
+	updateReturnChan chan struct{}
+}
+
+func NewMetricScraper(ctx context.Context, cluster *state.Cluster) *MetricScraper {
+	mc := &MetricScraper{
+		Cluster:          cluster,
+		terminateChan:    make(chan struct{}),
+		updateChan:       make(chan struct{}),
+		updateReturnChan: make(chan struct{}),
+	}
+	mc.init(ctx)
+	return mc
+}
+
+func (ms *MetricScraper) Terminate() {
```

Comment:
I'm not sure we want to expose an explicit terminate method. Other similar mechanisms function by stopping when ctx.Done().